### DISCUSSION TITLE
Add a rug feature

### DISCUSSION THREAD

**Discussion by davystrong** [PyO3/pyo3#3691] [Ideas]
I wrote some code implementing prime factorisation in Rust interfacing with Python. I initially used the `num-bigint` library supported by PyO3, but later discovered it seems to be fairly slow. I then switched to using `rug` (which uses GMP behind the scenes) which seems to be quite a bit faster. I wrote an additional feature flag for PyO3 to enable conversion, basically copied from the existing `bigint` implementation. Should I open a pull request or is this too like something which already exists? One issue to be aware of is that GMP (and thus `rug`) is licensed under LGPL, and I'm not sure if this would infect PyO3, or how that would work.

**Comment by mejrs** [comment]
A pull request is welcome 🙂 . That said I don't know about the licensing.

**Comment by davystrong** [comment]
Hmm, it appears that the license is an issue: LGPL requires any non (L)GPL calling code to allow for dynamic linking, which isn't ideal. I'm going to try to work out if I can write a package to allow overriding a normal Rust import with a dynamic library, as this should solve the licensing issue without compromising much on usability or performance. Don't know if you want to close this discussion, up to you.

**Reply by adamreichold** [reply]
I think one point here is that since we are also Apache licensed which is compatible with the GPL, I think we could have this single module be LGPL. So if the feature is disabled, PyO3 stays Apache OR MIT whereas if the feature is enabled it becomes essentially LGPL (subsuming the rest of the code base as a Apache-licensed).

I guess this would mainly be a problem with documenting this inside and around the code base. But personally, I think this might actually be preferable to playing tricks with the loader to try to work around the intention of the license.

**Comment by ms314159** [comment]
I was wondering if there has been any progress on this issue. It would be a very useful feature to have. Thank you.

**Reply by davystrong** [reply]
After discovering the issue with LGPL I more or less gave up on this. I don't see much point in working on a *GPL library.

**Reply by ms314159** [reply]
@adamreichold @mejrs Would the PyO3 team consider implementing this feature?

**Reply by davystrong** [reply]
@ms314159 I have published my code [here](https://github.com/davystrong/pyo3/tree/feature-rug) if you're interested, on the `feature-rug` branch. Feel free to use it as you wish. It should be working, but I don't remember if I included any example code.
